CI: skip tests when only files in doc/web changes (github actions)

ShaharNaveh opened the pull request:
- [x] closes #41101
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them

Commit:
CI: skip tests when only files in doc/web changes (github)

ShaharNaveh commented on .github/workflows/ci.yml:
```diff
@@ -99,14 +99,29 @@ jobs:
+    # Github workflows built-in path filters
+    # don't work on a level of individual jobs or steps.
+    # So, we use this.
+    - uses: dorny/paths-filter@v2
```
I really did not want to add an external "action", but as the comment says, github action don't work on the individual step level, only at the whole workflow.

So, either that this stays like that. Or we will have a separate workflow file for this.
I don't mind doing it either way, just LMK what is preferred :)

ShaharNaveh commented on .github/workflows/ci.yml:
```diff
@@ -123,14 +138,15 @@ jobs:
     - name: Upload web
       run: rsync -az --delete --exclude='pandas-docs' --exclude='docs' --exclude='Pandas_Cheat_Sheet*' web/build/ docs@${{ secrets.server_ip }}:/usr/share/nginx/pandas
-      if: github.event_name == 'push'
+      if: ${{ github.event_name == 'push' && steps.changes.outputs.web == 'true' }}
```
The reason for this edit is that I am afraid that because there is a way that we skip the build, so ```rsync``` will sync the local directory with the remote directory and if the local directory is empty then it might delete everything at the server?

I'm not sure, so I added this, will remove if I understood the command wrong.

Commit:
Fixed yaml file indent

jorisvandenbossche commented:
@ShaharNaveh Looking at the diff, I think this PR is making it so that we only do the doc biuld if anything in `doc/**` or `web/**` is touched, is that a correct understanding?

However, I think we actually want to other way around: we can skip the other (non-doc) builds if *only* `doc/**` is touched.

The doc build also needs to run if eg a docstring has changed in the actual pandas source.

ShaharNaveh commented:
> @ShaharNaveh Looking at the diff, I think this PR is making it so that we only do the doc biuld if anything in `doc/**` or `web/**` is touched, is that a correct understanding?

That's correct.

> However, I think we actually want to other way around: we can skip the other (non-doc) builds if _only_ `doc/**` is touched.

Oh my, I completly understood that wrong :sweat_smile:

Commit:
Merge remote-tracking branch 'upstream/master' into CI-action-on-change

Commit:
Revert "CI: skip tests when only files in doc/web changes (github)"

This reverts commit 7eaca3fdc3f2e98f2bc9748b37d8ab86d4f6ee2a.

Commit:
CI: Don't run lint if only doc changes (github worklow)

Commit:
CI: Don't run lint if only doc changes (azure)

Commit:
Merge remote-tracking branch 'upstream/master' into CI-action-on-change

Commit:
Fix (ex|in)clude logic in azure templates

Commit:
Fixed small "typo"

Commit:
Fix location of the "path" parameter

Commit:
Removed extra "trigger"

Commit:
Merge remote-tracking branch 'upstream/master' into CI-action-on-change

datapythonista commented on azure-pipelines.yml:
```diff
@@ -1,7 +1,14 @@
+    - master
+    - 1.2.x
+  paths:
+    include:
```
Is this `include` needed? Feels like his should be the default.

datapythonista reviewed (commented):
Looks good., just couple of things.

@jorisvandenbossche do you want to skip the tests in the master builds too? Feels like only skipping for PRs, and running full master builds can make sense too.

@ShaharNaveh do you think you can commit this in the master of your fork, and open a PR in your fork modifying the docs, and another modifying the docs and something else, and see if he builds run as expected? I think should be quite fast to set up, and we'll be on the right side, and avoid much disruption if we're missing something here. Thanks!

jorisvandenbossche commented:
> do you want to skip the tests in the master builds too? Feels like only skipping for PRs, and running full master builds can make sense too.

Yes, I think it would be good to only skip the tests on PRs when appropriate, and on master still always run the tests.

jorisvandenbossche commented on .github/workflows/ci.yml:
```diff
@@ -3,10 +3,14 @@ name: CI
       - master
       - 1.2.x
+    paths-ignore:
+      - "doc/**"
```
For this file, we probably don't want to include it (for all others it's fine). Because this file contains the doc build, and we still want to run that if there are doc changes.

Commit:
Merge remote-tracking branch 'upstream/master' into CI-action-on-change

Commit:
Fix for comment

https://github.com/pandas-dev/pandas/pull/41310/files#r629288230

Commit:
Fix for comment

https://github.com/pandas-dev/pandas/pull/41310/files#r629289631

Commit:
Fix for comment

https://github.com/pandas-dev/pandas/pull/41310/files#r628420098

ShaharNaveh commented:
> @ShaharNaveh do you think you can commit this in the master of your fork, and open a PR in your fork modifying the docs, and another modifying the docs and something else, and see if he builds run as expected? I think should be quite fast to set up, and we'll be on the right side, and avoid much disruption if we're missing something here. Thanks!


Sure:)

ShaharNaveh commented:
> > @ShaharNaveh do you think you can commit this in the master of your fork, and open a PR in your fork modifying the docs, and another modifying the docs and something else, and see if he builds run as expected? I think should be quite fast to set up, and we'll be on the right side, and avoid much disruption if we're missing something here. Thanks!
> 
> Sure:)

https://github.com/ShaharNaveh/pandas/pull/1
https://github.com/ShaharNaveh/pandas/pull/2

ShaharNaveh commented:
> > > @ShaharNaveh do you think you can commit this in the master of your fork, and open a PR in your fork modifying the docs, and another modifying the docs and something else, and see if he builds run as expected? I think should be quite fast to set up, and we'll be on the right side, and avoid much disruption if we're missing something here. Thanks!
> > 
> > 
> > Sure:)
> 
> [ShaharNaveh#1](https://github.com/ShaharNaveh/pandas/pull/1)
> [ShaharNaveh#2](https://github.com/ShaharNaveh/pandas/pull/2)



> > > @ShaharNaveh do you think you can commit this in the master of your fork, and open a PR in your fork modifying the docs, and another modifying the docs and something else, and see if he builds run as expected? I think should be quite fast to set up, and we'll be on the right side, and avoid much disruption if we're missing something here. Thanks!
> > 
> > 
> > Sure:)
> 
> [ShaharNaveh#1](https://github.com/ShaharNaveh/pandas/pull/1)
> [ShaharNaveh#2](https://github.com/ShaharNaveh/pandas/pull/2)

One thing that should be noted is that in PR#2 (Doc and Code changes) that the Azure pipeline does not trigger at all, but I think this is normal (I hope)

jorisvandenbossche commented:
Thanks for testing it!

> One thing that should be noted is that in PR#2 (Doc and Code changes) that the Azure pipeline does not trigger at all, but I think this is normal (I hope)

Yes, I think that is indeed to be expected, since AFAIK Azure Pipelines are not activated on forks

jorisvandenbossche commented on .github/workflows/ci.yml:
```diff
@@ -3,6 +3,8 @@ name: CI
   push:
     branches: [master]
+    paths-ignore:
+      - "doc/**"
```
Can you remove the one for master here as well?

jorisvandenbossche commented on .github/workflows/database.yml:
```diff
@@ -3,10 +3,14 @@ name: Database
   push:
     branches: [master]
+    paths-ignore:
+      - "doc/**"
```
Let's only keep the ignore for "pull_request", so you can remove this one for "push" to master

jorisvandenbossche commented on .github/workflows/posix.yml:
```diff
@@ -3,10 +3,14 @@ name: Posix
   push:
     branches: [master]
+    paths-ignore:
+      - "doc/**"
```
Same here, and the ones below as well

Commit:
Fix for comment

https://github.com/pandas-dev/pandas/pull/41310#discussion_r630998670

Commit:
Fix for comment

https://github.com/pandas-dev/pandas/pull/41310#discussion_r630999072

Commit:
Fix for comment

https://github.com/pandas-dev/pandas/pull/41310#discussion_r630999256

Commit:
Only ignore on "pull_request" not on "push"

datapythonista commented:
@ShaharNaveh do you have time to address the last comments please? Thanks!

Commit:
Merge remote-tracking branch 'upstream/master' into CI-action-on-change

ShaharNaveh commented:
> @ShaharNaveh do you have time to address the last comments please? Thanks!

I addressed those comments a while back, just forgot to mark them as "resolved".
Can you please review this PR again @jorisvandenbossche ?

MarcoGorelli commented on .github/workflows/pre-commit.yml:
```diff
@@ -2,6 +2,8 @@ name: pre-commit
 
 on:
   pull_request:
+    paths-ignore:
+      - "doc/**"
```
This might as well always run, it's fast and has some checks for .rst files

MarcoGorelli reviewed (changes requested):
Great idea! Just got a comment

Commit:
Addresing comments

Ref: https://github.com/pandas-dev/pandas/pull/41310#discussion_r644939668

Commit:
Merge remote-tracking branch 'upstream/master' into CI-action-on-change

MarcoGorelli commented on azure-pipelines.yml:
```diff
@@ -1,7 +1,12 @@
+    - 1.2.x
+  paths:
+    exclude:
+    - 'doc/*'
```
any reason why the other have `doc/**` and this one has `doc/*`?

ShaharNaveh replied:
> any reason why the other have `doc/**` and this one has `doc/*`?

Yes, because this is the "azure" CI, and the other files are related to the "github actions" CI.
I just copied the syntax from their examples in the docs, I haven't really tested the differences.

MarcoGorelli reviewed (changes requested):
Thanks @ShaharNaveh - this looks correct to me, and it can be easily reverted if there's any issues, so let's merge and see

Commit:
Merge remote-tracking branch 'upstream/master' into CI-action-on-change

MarcoGorelli reviewed (approved):
Thanks @ShaharNaveh - this looks correct, and if it's not it's easy to revert, so let's merge

simonjayhawkins commented:
@MarcoGorelli our release scripts work from master and not a commit hash. will try to push the tag to master but may fail. in that case will need to change the milestone on this and the other PR.
